addons/website_sale_delivery/controllers/main.py

```diff
@@ -46,9 +47,188 @@ def cart_carrier_rate_shipment(self, carrier_id, **kw):
+
+        if order.partner_shipping_id.is_public:
+            shipping_address['name'] = _(
+                'Anonymous express checkout partner for order %s', order.name
+            )
+            order.partner_id = self._create_or_edit_partner(
+                shipping_address, type='delivery'
+            )
+            order.env.remove_to_compute(order.env['sale.order']._fields['pricelist_id'], order)
+        elif order.partner_shipping_id.name.endswith(order.name):
+            self._create_or_edit_partner(
+                shipping_address,
+                edit=True,
+                type='delivery',
+                partner_id=order.partner_shipping_id.id,
+            )
+        elif all(shipping_address[k] == order.partner_id[k] for k in shipping_address):
+            # If the user is logged in and the shipping address is the same as the billing address,
+            # we use the billing address
+            order.partner_shipping_id = order.partner_id
+        else:
+            shipping_partners_sudo = request.env['res.partner'].with_context(
+                show_address=1
+            ).sudo().search([
+                ('id', 'child_of', order.partner_id.commercial_partner_id.ids),
+            ])
+            order.partner_shipping_id = False  # Remove the predefined shipping partner id
+            for shipping_partner_sudo in shipping_partners_sudo:
+                # Check if a child partner doesn't already exist with the same informations. The
+                # phone isn't always checked because it isn't sent in shipping information with
+                # Google Pay
+                if all(shipping_address[k] == shipping_partner_sudo[k] for k in shipping_address):
+                    order.partner_shipping_id = shipping_partner_sudo.id
+                    break
+            if not order.partner_shipping_id:
+                # If there is no partner shpping id on the order, it means that no delivery partner
+                # was found with the same information so we creates a new one.
+                shipping_address['name'] = _(
+                    'Anonymous express checkout partner for order %s', order.name
+                )
+                order.partner_shipping_id = self._create_or_edit_partner(
+                    shipping_address, type='delivery', parent_id=order.partner_id.id,
+                )
+
+        # Return the list of develivery carrier available for the sale order.
+        return sorted([{
+            'id': carrier.id,
+            'name': carrier.name,
+            'description': carrier.website_description,
+            'minorAmount': payment_utils.to_minor_currency_units(
+                self._get_rate(carrier, order, is_express_checkout_flow=True)['price'],
+                order.currency_id,
+            ),
+        } for carrier in order._get_delivery_methods()], key=lambda carrier: carrier['minorAmount'])
+
+    def _get_rate(self, carrier, order, is_express_checkout_flow=False):
```

Comment:
Might be useful to tag it as a static method for your tests?

Reply:
PRO: No need to instantiate objects in tests